docs/use_cases/scaling_rag_for_production.md

```diff
@@ -0,0 +1,641 @@
+3. Retrieve the most relevant document chunks, determined by the proximity of the query vector embedding to the embedded document chunks.
+4. Add the retrieved document chunks as context to the query vector embedding and send it to the LLM.
+5. The LLM generates a response utilizing the retrieved context.
+
+According to this workflow, the following components are required: an embedding model, a store for document and vector embeddings, a retriever, and a LLM. While RAG workflows can become significantly more complex, incorporating methods like metadata filtering and retrieval reranking, it’s essential to first establish a strong foundation with these basic elements.
+
+## The LangChain question
+
+LangChain has arguably become the most prominent LLM library to this date. A lot of developers are using it to build Proof-of-Concepts (PoC) and Minimal Viable Products (MVPs) or to simply test new ideas. While there has been a lot of discussion about LangChain in production, *most* of the criticism can be boiled down to personal preference and the fact that LangChain was originally built to address problems occurring much earlier in the development cycle.
+
+So what to do? Keep in mind that this is merely my personal opinion since there are no gold standards for which tools to use yet, but I’m convinced that there is no universal answer to this question. All of the major LLM and RAG libraries - LangChain, LlamaIndex and Haystack, to name my personal top three - have what it takes to producitonize a RAG system. And there’s a simple reason for this: they all have integrations for third party libraries and providers that will handle the production requirements. I would try to view these tools as interfaces between all the other components. Which one you’d want to choose will depend on the details of your existing tech stack and use case.
+
+## The right tools for this tutorial
+
+Alright, but what will *we* choose for this tutorial? One of the first decisions to undertake will be where we want to run our system: do we want to use a cloud service or do we want to run it within our own network? Because I think tutorials should try to reduce complexity and avoid proprietary solutions where possible, we won’t choose the cloud option here. while the above mentioned libraries all support cloud deployment for AWS, Azure and GCP, the details of putting a RAG system in production wil heavily depend on the cloud provider of your choice. Instead, we will utilize [Ray](https://github.com/ray-project/ray).
+
+Ray is a Python framework for productionizing and scaling ML workloads. It's adaptable for both local environments and Kubernetes, efficiently managing all workload requirements. While we could opt for Ray integrations like LangChain, LlamaIndex, or Haystack, it's worth considering using Ray directly. This approach might provide more universally applicable insights, given that these integrations are all built upon the same underlying tool.

```

Comment:
I think it is not clear why we need Ray if we don't do cloud. I think is is worth explaining in a sentence at least.

Reply:
Added a sentence in the new "producitonizing" paragraph.